### PR TITLE
feat: use v2 API for agent lifecycle updates

### PR DESCRIPTION
**Pull request by spikecurtis**
Agent uses the v2 API to post lifecycle updates.

Part of #10534

### CODE REVIEW

**Comment by spikecurtis** [comment]
* **#12278** <a href="https://app.graphite.dev/github/pr/coder/coder/12278?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @spikecurtis and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: agent/agent.go [hunk 1]
```diff
@@ -644,33 +640,35 @@ func (a *agent) reportLifecycleUntilClose() {
...
 				break
 			}
+			l, err := agentsdk.ProtoFromLifecycle(report)
+			if err != nil {
+				a.logger.Critical(ctx, "failed to convert lifecycle state", slog.F("report", report))
+				// Skip this report; there is no point retrying.  Maybe we can successfully convert the next one?
+				lastReportedIndex++
+				continue
+			}
+			logger := a.logger.With(slog.F("payload", l))
```
Comment (mafredri) [inline agent/agent.go:650]: Nit: The `payload` should be `payload := &proto.UpdateLifecycleRequest{Lifecycle: l}`, as that's what's being sent.

**Review by mafredri — APPROVED** [review]
Some tests need updating (relying on removed `agentClient.PostLifecycle`), but otherwise LGTM!

**Comment by spikecurtis** [comment]
Thanks! Going to keep the PostLifecycle around, but deprecated so that we have tests against the v1 API.

We can retire it all when we drop v1 support.

**Comment by spikecurtis** [comment]
### Merge activity

* **Feb 23, 6:24 AM EST**: @spikecurtis merged this pull request with [Graphite](https://app.graphite.dev/github/pr/coder/coder/12278).

**Pull request merged**